azurerm/internal/services/signalr/signalr_service_resource.go

```diff
@@ -390,6 +465,57 @@ func flattenSignalRFeatures(features *[]signalr.Feature) []interface{} {
 	return result
 }
 
+func expandUpstreamSettings(input []interface{}) *signalr.ServerlessUpstreamSettings {
+	upstreamTemplates := make([]signalr.UpstreamTemplate, 0)
+
+	for _, upstreamSetting := range input {
+		setting := upstreamSetting.(map[string]interface{})
+
+		upstreamTemplate := &signalr.UpstreamTemplate{
+			HubPattern:      utils.String(strings.Join(*utils.ExpandStringSlice(setting["hub_pattern"].([]interface{})), ",")),
+			EventPattern:    utils.String(strings.Join(*utils.ExpandStringSlice(setting["event_pattern"].([]interface{})), ",")),
+			CategoryPattern: utils.String(strings.Join(*utils.ExpandStringSlice(setting["category_pattern"].([]interface{})), ",")),
+			URLTemplate:     utils.String(setting["url_template"].(string)),
+		}
+
+		upstreamTemplates = append(upstreamTemplates, *upstreamTemplate)
+	}
+
+	return &signalr.ServerlessUpstreamSettings{
+		Templates: &upstreamTemplates,
+	}
+}
+
+func flattenUpstreamSettings(upstreamSettings *signalr.ServerlessUpstreamSettings) []interface{} {
+	result := make([]interface{}, 0)
+	if upstreamSettings == nil || upstreamSettings.Templates == nil {
+		return result
+	}
+
+	for _, settings := range *upstreamSettings.Templates {
+		upstreamSetting := make(map[string]interface{})
+
+		if settings.CategoryPattern != nil {
+			categoryPatterns := strings.Split(*settings.CategoryPattern, ",")
+			upstreamSetting["category_pattern"] = utils.FlattenStringSlice(&categoryPatterns)
+		}
+		if settings.EventPattern != nil {
+			eventPatterns := strings.Split(*settings.EventPattern, ",")
+			upstreamSetting["event_pattern"] = utils.FlattenStringSlice(&eventPatterns)
+		}
+		if settings.HubPattern != nil {
+			hubPatterns := strings.Split(*settings.HubPattern, ",")
+			upstreamSetting["hub_pattern"] = utils.FlattenStringSlice(&hubPatterns)
+		}
+		if settings.URLTemplate != nil {
+			upstreamSetting["url_template"] = *settings.URLTemplate
+		}
+
+		result = append(result, upstreamSetting)
```

Comment:
each of these properties needs to be set to a value even if the API is nil, so that we can track changes here - so can we make this:

```suggestion
 		urlTemplate := ""
 		if settings.URLTemplate != nil {
 			urlTemplate = *settings.URLTemplate
 		}
 		hubPattern := make([]string)
 		if settings.HubPattern != nil {
 			hubPatterns := strings.Split(*settings.HubPattern, ",")
 			hubPattern = utils.FlattenStringSlice(&hubPatterns)
 		}
 		
 		result = append(result, map[string]interface{}{
 			"url_template": urlTemplate,
 		})
```

Reply:
To make sure I'm fully understanding this statement. By "we can track changes here" you mean so that the changes can be tracked in TF State @tombuildsstuff ?

Reply:
missed this - but yes